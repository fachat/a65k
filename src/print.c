/****************************************************************************

    print output
    Copyright (C) 2012,2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <stdarg.h>
#include <ctype.h>
#include <string.h>
#include <errno.h>

#include "print-config.h"
#include "print.h"
#include "arith.h"

// filler spaces
static const char spaces[] = "                                        ";

// column widths for fixed width output, index corresponds to printer_col_t
static const int widths[] = { 
	6, 	// PRT_LINENO
	5,	// PRT_ADDR
	9,	// PRT_DATA
	11,	// PRT_LABEL
	4,	// PRT_OPERATION
	20,	// PRT_PARAM
	30	// PRT_COMMENT
};

void print_module_init() {

	print_config_init();
}

static type_t printer_memtype = {
        "printer_t",
        sizeof(printer_t)
};

// initialize printer output channel
printer_t *print_init() {

	print_config_t *cfg = print_current_config();

	const char *filename = cfg->filename;
	FILE *f = NULL;
	if (filename != NULL && strlen(filename) > 0) {
		f = fopen(filename, "w");
		if (f == NULL) {
			log_error("Could not open output file '%s' -> %s", filename, strerror(errno));
			return NULL;
		}
	}

	printer_t *prt = mem_alloc(&printer_memtype);

	prt->outfile = f;
	prt->cfg = cfg;
	prt->current_col = PRT_LINENO;

	int p = 0;
	for (int i = PRT_LINENO; i < PRT_END; i++) {
		prt->col_width[i] = p;
		p += widths[i];
	}

	print_clr(prt);

	prt->was_empty = 0;

	return prt;
}

static int print_getlen(printer_t *prt) {
	return prt->buflen;
}

static void print_setcol(printer_t *prt, int col) {
	
	int buflen = print_getlen(prt);

	int todo = col - buflen;
	int slen = strlen(spaces);

	if (todo <= 0) {
		todo = 1;
	}

	while (todo > 0) {
		int step = slen - todo;
		if (step < 0) {
			step = 0;
		}
		print(prt, prt->current_col, "%s", spaces + step);
		todo -= slen + step;
	}
}


void print(printer_t *prt, print_col_t col_no, const char *pattern, ...) {

	if (col_no != prt->current_col) {
		// target fixed width column
		if (col_no == PRT_COMMENT && prt->current_col == PRT_LINENO) {
			// shift comment to left if only comment on line
			col_no = PRT_LABEL;
		}
		int trg = prt->col_width[col_no];
		print_setcol(prt, trg);
		prt->current_col = col_no;
	}

        va_list ap;
        va_start(ap, pattern);

	int buflen = print_getlen(prt);

        int r = vsnprintf(prt->buf + buflen, BUF_LEN - buflen, pattern, ap);
        if (r < 0 || r > BUF_LEN) {
                // error
                log_error("Error printing %d\n", r);
                return;
        }
	prt->buflen += r;
}

void print_out(printer_t *prt) {

	if (prt->buflen == 0) {
		if (prt->cfg->collapse_empty && prt->was_empty) {
			return;
		}
		prt->was_empty = 1;
	} else {
		prt->was_empty = 0;
	}

	if (prt->outfile == NULL) {
        	printf("%s\n", prt->buf);
	} else {
        	fprintf(prt->outfile, "%s\n", prt->buf);
	}
	print_clr(prt);
}

void print_clr(printer_t *prt) {
	prt->buf[0] = 0;
	prt->buflen = 0;
	prt->current_col = PRT_LINENO;
}

void print_close(printer_t *prt) {

	if (prt->outfile) {
		fclose(prt->outfile);
	}
}


