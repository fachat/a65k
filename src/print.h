/****************************************************************************

    print output
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef PRINT_H
#define PRINT_H

#include <stdio.h>

#include "parser.h"
#include "print-config.h"

#define	BUF_LEN		4096

// definitions for print output columns
typedef enum {
	PRT_LINENO		= 0,		// line number
	PRT_ADDR		= 1,		// address of line
	PRT_DATA		= 2,		// bytes generated by line
	PRT_LABEL		= 3,		// label definition
	PRT_OPERATION		= 4,		// operation
	PRT_PARAM		= 5,		// parameter
	PRT_COMMENT		= 6,		// comment
	PRT_END			= 7		// unused, defined for len
} print_col_t;

typedef struct {
	FILE			*outfile;
	const char		*filename;
	const print_config_t	*cfg;
	// output buffer
	char			buf[BUF_LEN];
	int			buflen;
	// runtime flags
	// last line was empty
	int			was_empty;
	// column widths for fixed column width
	print_col_t		current_col;
	int			col_width[PRT_END];
} printer_t;

// init print module
void print_module_init();

// initialize printer output channel
printer_t *print_init();

// general print; 
// Note: switching col_no automatically goes to this column, but
// only in ascending order. Printing with same col_no is allowed
// and simply appends.
void print(printer_t *ptr, print_col_t col_no, const char *pattern, ...);

// print a statement formatted
void print_formatted_stmt(printer_t *cfg, const statement_t *stmt);

void print_out(printer_t *prt);

void print_clr(printer_t *prt);

void print_close(printer_t *prt);

#endif

