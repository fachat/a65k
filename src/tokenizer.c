/****************************************************************************

    tokenizing a line
    Copyright (C) 2015 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

// this code simply tokenizes the input string given some simple basic rules.

#include <ctype.h>
#include <stdbool.h>
#include <string.h>

#include "types.h"
#include "mem.h"
#include "err.h"
#include "tokenizer.h"
#include "log.h"


static type_t tokenizer_memtype = {
	"tokenizer_t",
	sizeof(tokenizer_t)
};

static op_details_t *sorted[256];

static op_details_t ops[] = {
        { 	OP_EXCL,		"!",	1,	0,	0,	0,	0	},
        {	OP_AT,			"@",	1,	0,	0,	0,	0	},
        {	OP_XQUOTE,		"`",	0,	0,	0,	0,	0	},
        {	OP_SQUOTE,		"'",	0,	0,	0,	0,	0	},
        {	OP_OPEN,		"(",	0,	0,	0,	0,	0	},
        {	OP_CLOSE,		")",	0,	0,	0,	0,	0	},
        {	OP_HASH, 		"#",	0,	0,	0,	0,	0	},
        {	OP_BOPEN,		"[",	0,	0,	0,	0,	0	},
        {	OP_BBOPEN,		"[[",	0,	0,	0,	0,	0	},
        {	OP_BCLOSE,		"]",	0,	0,	0,	0,	0	},
        {	OP_BBCLOSE,		"]]",	0,	0,	0,	0,	0	}, 
        {	OP_STAR,		"*",	0,	0,	1,	0,	0	},
        {	OP_COMMA,		",",	0,	0,	0,	0,	0	},
        {	OP_PLUS,		"+",	0,	0,	1,	0,	0	},
        {	OP_MINUS,		"-",	0,	1,	1,	0,	0	},
        {	OP_DIV,			"/",	0,	0,	1,	0,	0	},
        {	OP_COLON,		":",	0,	0,	0,	0,	0	},
        {	OP_SEMICOLON,		";",	0,	0,	0,	0,	0	},
        {	OP_DOT,			".",	0,	0,	0,	0,	0	},
        {	OP_LARGER,		">",	1,	0,	1,	0,	0	},
        {	OP_LARGEROREQUAL, 	">=",	1,	0,	1,	0,	0	},
        {	OP_LESS,		"<",	1,	0,	1,	0,	0	},
        {	OP_LESSOREQUAL,		"<=",	0,	0,	1,	0,	0	},
        {	OP_ASSIGN,		"=",	0,	0,	0,	0,	1	}, 
        {	OP_ASSIGNADD,		"+=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNSUB,		"-=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNMULT,		"*=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNDIV,		"/=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNMOD,		"%=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNBITOR,		"|=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNBITAND,	"&=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNBITXOR,	"^=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNSHIFTLEFT,	"<<=",	0,	0,	0,	0,	1	},
        {	OP_ASSIGNSHIFTRIGHT,	">>=",	0,	0,	0,	0,	1	},
        {	OP_BITXOR,		"^",	1,	0,	1,	0,	0	},
        {	OP_BITOR,		"|",	0,	0,	1,	0,	0	},
        {	OP_BITAND,		"&",	0,	0,	1,	0,	0	},
        {	OP_EQUAL,		"==",	0,	0,	1,	0,	0	},
        {	OP_NOTEQUAL,		"!=",	0,	0,	1,	0,	0	},
        {	OP_SHIFTLEFT,		"<<",	0,	0,	1,	0,	0	},
        {	OP_SHIFTRIGHT,		">>",	0,	0,	1,	0,	0	},
        {	OP_LOGICAND,		"&&",	0,	0,	1,	0,	0	},
        {	OP_LOGICOR,		"||",	0,	0,	1,	0,	0	},
        {	OP_LOGICXOR,		"^^",	0,	0,	1,	0,	0	},
        {	OP_BITINV,		"~",	1,	0,	1,	0,	0	},
        {	OP_MOD,			"%",	1,	0,	1,	0,	0	},
        {	OP_DOUBLESLASH,		"//",	0,	0,	0,	0,	0	},
        {	OP_XIND,		",x",	0,	0,	0,	1,	0	},
        {	OP_YIND,		",y",	0,	0,	0,	1,	0	},
        {	OP_ZIND,		",z",	0,	0,	0,	1,	0	},
        {	OP_SIND,		",s",	0,	0,	0,	1,	0	},
        {	OP_BIND,		",b",	0,	0,	0,	1,	0	},
        {	OP_PIND,		",p",	0,	0,	0,	1,	0	},
};

void tokenizer_module_init(void) {

	memset(sorted, 0, 256 * sizeof(op_details_t*));

	unsigned int n = sizeof(ops) / sizeof(op_details_t);

	for (unsigned int i = 0; i < n; i++) {
		op_details_t *p = &ops[i];	

		sorted[p->type] = p;
	}
}


op_details_t *tokenizer_op_details(op_t op) {
	op_details_t *od = sorted[op];

	if (od == NULL) {
		log_fatal("illegal token %d\n", op);
	}
	return od;
}



// initialize a tokenizer 
tokenizer_t *tokenizer_create(const char *line, int cstyle_allowed) {
	
	tokenizer_t *tok = mem_alloc(&tokenizer_memtype);

	tok->line = line;
	tok->cstyle_allowed = cstyle_allowed;
	tok->ptr = 0;
	tok->len = 0;
	tok->type = T_INIT;

	return tok;
}

static bool_t parse_base(tokenizer_t *tok, int ptr, int base) {

	const char *line = tok->line;
	signed char c = 0;

	long value = 0;
	while (line[ptr] != 0) {
		c = tok->line[ptr];
		if (!isalnum(c)) {
			// other tokens like ",", or operators
			break;
		}
		if (isdigit(c)) {
			c &= 0x0f;
		} else {
			c = tolower(c) - 'a' + 10;
		}
		if (c > (base-1)) {
			// TODO: continue till end of number, so simple typo still allows detect further errors
			tok->vals.errno = E_TOK_DIGITRANGE;
			tok->type = T_ERROR;
			return false;
		}
		value = value * base + c;
		ptr++;
	}
	tok->vals.literal.value = value;
	tok->len = ptr - tok->ptr;

	if (tok->len == 0) {
		// syntax error
		tok->vals.errno = E_TOK_EMPTY;
		tok->type = T_ERROR;
		return false;
	}
	return true;
}

static inline bool_t parse_decimal(tokenizer_t *tok, int ptr) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = LIT_DECIMAL;
	return parse_base(tok, ptr, 10);
}

static inline bool_t parse_octal(tokenizer_t *tok, int ptr, int is_c) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = is_c ? LIT_OCTAL_C : LIT_OCTAL;
	return parse_base(tok, ptr, 8);
}

static inline bool_t parse_binary(tokenizer_t *tok, int ptr) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = LIT_BINARY;
	return parse_base(tok, ptr, 2);
}

static inline bool_t parse_hex(tokenizer_t *tok, int ptr, int is_c) {

	tok->type = T_LITERAL;
	tok->vals.literal.type = is_c ? LIT_HEX_C : LIT_HEX;
	return parse_base(tok, ptr, 16);
}

static inline bool_t parse_string(tokenizer_t *tok, int ptr) {

	const char *line = tok->line;

	char delim = line[ptr];
	tok->type = T_STRING;
	tok->vals.string.type = delim;

	ptr++;
	tok->vals.string.ptr = ptr;
	tok->vals.string.len = 0;

	while (isPrint(line[ptr])) {
		if (line[ptr] == delim) {
			break;
		}
		ptr++;
	}
	tok->vals.string.len = ptr - tok->vals.string.ptr;
	
	if (line[ptr] == delim) {
		ptr++;
	} else 
	if (line[ptr] != 0) {
		// non-printable char ends string
		tok->vals.errno = E_TOK_NONPRINT;
		tok->type = T_ERROR;
		return false;
	}

	tok->len = ptr - tok->ptr;

	if (tok->vals.string.len == 0) {
		// empty string
		tok->vals.errno = E_TOK_EMPTY;
		tok->type = T_ERROR;
		return false;
	}
	return true;
}

static inline bool_t parse_name(tokenizer_t *tok, int ptr) {

	const char *line = tok->line;
	tok->type = T_NAME;

	while (isalnum(line[ptr]) || (line[ptr] == '_')) {
		ptr++;
	}

	tok->len = ptr - tok->ptr;

	return true;
}

static inline bool_t parse_token(tokenizer_t *tok, int ptr, int can_have_operator, int allow_index) {

	const char *line = tok->line;

	char c = line[ptr];

	// default outcome
	tok->vals.op = c;
	tok->type = T_TOKEN;
	tok->len = 1;

	switch(c) {
	case '@':
	case '`':
	case '\'':
	case '#':
	case ':':
	case '.':
	case ';':
		return true;
	case '(':
	case ')':
		tok->type = T_BRACKET;
		return true;
	case ',':
		if (allow_index) {
			// check OP_*IND
			char c = tolower(line[ptr+1]);
			if (c == 'x') {
				tok->len++;
				tok->vals.op = OP_XIND;
			} else
			if (c == 'y') {
				tok->len++;
				tok->vals.op = OP_YIND;
			} else
			if (c == 'z') {
				tok->len++;
				tok->vals.op = OP_ZIND;
			} else
			if (c == 'b') {
				tok->len++;
				tok->vals.op = OP_BIND;
			} else
			if (c == 's') {
				tok->len++;
				tok->vals.op = OP_SIND;
			} else
			if (c == 'p') {
				tok->len++;
				tok->vals.op = OP_PIND;
			}
		}
		return true;
	case '%':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNMOD;
			tok->len++;
			return true;
		}
		return true;
	case '-':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNSUB;
			tok->len++;
			return true;
		}
		return true;
	case '+':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNADD;
			tok->len++;
			return true;
		}
		return true;
	case '*':
		if (line[ptr+1] == '=') {
			tok->vals.op = OP_ASSIGNMULT;
			tok->len++;
			return true;
		}
		return true;
	case '/':
		switch (line[ptr+1]) {
		case '=':
			tok->vals.op = OP_ASSIGNDIV;
			tok->len++;
			return true;
		case '/':
			tok->vals.op = OP_DOUBLESLASH;
			tok->len++;
			return true;
		}
		return true;
	case ']':
		tok->type = T_BRACKET;
		if (line[ptr+1] == ']') {
			tok->vals.op = OP_BBCLOSE;
			tok->len++;
			return true;
		}
		// OP_BOPEN
		return true;
	case '[':
		tok->type = T_BRACKET;
		if (line[ptr+1] == '[') {
			tok->vals.op = OP_BBOPEN;
			tok->len++;
			return true;
		}
		// OP_BOPEN
		return true;
	case '!':
		if (can_have_operator) {
			switch (line[ptr+1]) {
			case '=':
				tok->vals.op = OP_NOTEQUAL;
				tok->len++;
				return true;
			default:
				// default
				// OP_EXCL
				return true;
			}
		} else {
			// OP_EXCL
			return true;
		}
	case '>':
		if (can_have_operator) {
			// check ">>", ">=", "><"
			switch (line[ptr+1]) {
			case '>':
				if (line[ptr+2] == '=') {
					tok->vals.op = OP_ASSIGNSHIFTRIGHT;
					tok->len += 2;
					return true;
				}
				tok->vals.op = OP_SHIFTRIGHT;
				tok->len++;
				return true;
			case '=':
				tok->vals.op = OP_LARGEROREQUAL;
				tok->len++;
				return true;
			case '<':
				tok->vals.op = OP_NOTEQUAL;
				tok->len++;
				return true;
			default:
				// default
				return true;
			}
		} else {
			// ">" for high byte modifier or larger than
			return true;
		}
	case '<':
		if (can_have_operator) {
			// check "<<", "<=", "<>"
			switch (line[ptr+1]) {
			case '<':
				if (line[ptr+2] == '=') {
					tok->vals.op = OP_ASSIGNSHIFTLEFT;
					tok->len += 2;
					return true;
				}
				tok->vals.op = OP_SHIFTLEFT;
				tok->len++;
				return true;
			case '=':
				tok->vals.op = OP_LESSOREQUAL;
				tok->len++;
				return true;
			case '>':
				tok->vals.op = OP_NOTEQUAL;
				tok->len++;
				return true;
			default:
				// default
				return true;
			}
		} else {
			// "<" for high byte modifier
			return true;
		}
	case '=':
		if (can_have_operator) {
			// check "==", "=>", "=<"
			switch (line[ptr+1]) {
			case '=':
				tok->vals.op = OP_EQUAL;
				tok->len++;
				return true;
			default:
				// default, assign
				return true;
			}
		} else {
			// '=' assign
			return true;
		}
	case '&':
		// check "&", "&&", "&="
		switch (line[ptr+1]) {
		case '&':
			tok->vals.op = OP_LOGICAND;
			tok->len++;
			return true;
		case '=':
			tok->vals.op = OP_ASSIGNBITAND;
			tok->len++;
			return true;
		} 
		return true;
	case '|':
		// check "|", "||", "|="
		switch (line[ptr+1]) {
		case '|':
			tok->vals.op = OP_LOGICOR;
			tok->len++;
			return true;
		case '=':
			tok->vals.op = OP_ASSIGNBITOR;
			tok->len++;
			return true;
		} 
		return true;
	case '^':
		// check "^", "^^", "^="
		switch (line[ptr+1]) {
		case '^':
			tok->vals.op = OP_LOGICXOR;
			tok->len++;
			return true;
		case '=':
			tok->vals.op = OP_ASSIGNBITXOR;
			tok->len++;
			return true;
		} 
		return true;
	default:
		tok->vals.errno = E_TOK_UNKNOWN;
		tok->type = T_ERROR;
		return false;
	}

	return true;
}

// set to next token; return true when there is a valid token
bool_t tokenizer_next(tokenizer_t *tok, int allow_index) {

	// move behind last token
	int ptr = tok->ptr + tok->len;

	const char *line = tok->line;

	// attempt to not have to store an extra operand/operator flag in the tokenizer	
	// or even have to pass it to this function as parameter...
	bool_t can_have_operator = (tok->type == T_LITERAL) 
			|| (tok->type == T_TOKEN && line[ptr] == ')');

	// skip whitespace
	while (line[ptr] != 0 && isspace(line[ptr])) {
		ptr++;
	}

	tok->ptr = ptr;

	//printf("ptr=%d, c=%d\n", ptr, line[ptr]);

	// are we done with the line yet?
	if (line[ptr] == 0 || tok->type == T_END || tok->type == T_ERROR) {
		tok->len = 0;
		if (tok->type != T_ERROR) {
			tok->type = T_END;
		}
		return false;
	}

	char c = line[ptr];

	if (isdigit(c)) {
		// handle C-style octal (starting with '0'), hex (starting with '0x'), dec
		if (tok->cstyle_allowed && c == '0') {
			if (line[ptr+1] == 'x' || line[ptr+1] == 'X') {
				// hex, C-style
				tok->ptr = ptr;
				ptr += 2;
				return parse_hex(tok, ptr, 1);
			} else {
				// octal
				tok->ptr = ptr;
				ptr += 1;
				return parse_octal(tok, ptr, 1);
			}
		} else {
			// dec
			return parse_decimal(tok, ptr);
		}
	} else
	if ((!can_have_operator) && (c == '%')) {
		// binary
		tok->ptr = ptr;
		ptr++;
		return parse_binary(tok, ptr);
	} else
	if ((!can_have_operator) && (c == '$')) {
		// hex
		tok->ptr = ptr;
		ptr++;
		return parse_hex(tok, ptr, 0);
	} else
	if ((!can_have_operator) && (c == '&')) {
		// octal
		ptr++;
		tok->ptr = ptr;
		return parse_octal(tok, ptr, 0);
	} else
	if (is_string_delim(c)) {
		// string literal
		// c must be of one of quotetype_t enum!
		return parse_string(tok, ptr);
	} else
	if (isalpha(c) || c == '_') {
		// name
		return parse_name(tok, ptr);
	} else
	if (ispunct(c)) {
		// any other token
		return parse_token(tok, ptr, can_have_operator, allow_index);
	} else {
		// non-printable - error
		tok->vals.errno = E_TOK_UNKNOWN;
		tok->type = T_ERROR;
		return false;
	}
	
}

// set to next token; return true when there is a valid name token, that is used as prefix
bool_t tokenizer_next_prefix(tokenizer_t *tok) {

	// move behind last token
	int ptr = tok->ptr + tok->len;

	const char *line = tok->line;

	// skip whitespace
	while (line[ptr] != 0 && isspace(line[ptr])) {
		ptr++;
	}

	tok->ptr = ptr;

	// are we done with the line yet?
	if (line[ptr] == 0 || tok->type == T_END || tok->type == T_ERROR) {
		tok->len = 0;
		if (tok->type != T_ERROR) {
			tok->type = T_END;
		}
		return false;
	}

	return parse_name(tok, ptr);
}

// set to next token; return true when there is a valid name token, that is used as prefix
bool_t tokenizer_next_comment(tokenizer_t *tok, int allow_colon) {

	// move behind last token
	int ptr = tok->ptr + tok->len;

	const char *line = tok->line;

	// skip whitespace
	while (line[ptr] != 0 && isspace(line[ptr])) {
		ptr++;
	}

	tok->ptr = ptr;
	tok->len = 0;

	// are we done with the line yet?
	if (line[ptr] == 0 || tok->type == T_END || tok->type == T_ERROR) {
		if (tok->type != T_ERROR) {
			tok->type = T_END;
		}
		return false;
	}

	tok->len = strlen(tok->line + ptr);
	if (allow_colon) {
		// TODO need escape handling?
		for (int i = 0; i < tok->len; i++) {
			if (line[tok->ptr + i] == ':') {
				tok->len = i;
			}
		}
	}

	return true;
}


void tokenizer_free(tokenizer_t *tok) {
	
	tok->line = NULL;

	mem_free(tok);
}



