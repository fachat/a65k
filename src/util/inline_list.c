/****************************************************************************

    list handling
    Copyright (C) 2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>

#include "types.h"
#include "inline_list.h"


static type_t ilist_memtype = {
	"ilist",
	sizeof(ilist_t)
};

ilist_t *inline_list_init(int bucket_size, const type_t *type, void (*init_item)(void*)) {

	ilist_t *ilist = mem_alloc(&ilist_memtype);

	ilist->cap = bucket_size;
	ilist->len = 0;
	ilist->type = type;
	ilist->init_item = init_item;
	ilist->values = mem_alloc_n(bucket_size, type);

	return ilist;
}


