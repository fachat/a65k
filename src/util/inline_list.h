/****************************************************************************

    list handling
    Copyright (C) 2017 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

/*
   A list that stores the entries in an array.
   (this is different from arraylist, that only stores pointers to the entry)

   Because this requires that a pointer to the memory for the object must
   be returned by the list, not added to it, the API changes and
   inline_list does not implement the general list interface.
*/


#ifndef INLINE_LIST_H
#define INLINE_LIST_H

#include <stdio.h>

#include "types.h"
#include "mem.h"


typedef struct {
	int		cap;		// capacity of list
	int		len;		// number of items in list
	int		popped;		// flag, if true add is not allowed anymore
	const type_t	*type;		// type info for array element
					// function to init a new item
	void 		(*init_item)(void*);
					// list of inlined items
	void		*values;
} ilist_t;

ilist_t *inline_list_init(int bucket_size, const type_t *type, void (*init_item)(void*));

static inline int ilist_len(const ilist_t *il) {
	return il->len;
}

static inline void* ilist_get_internal(const ilist_t *ilist, int itemno) {
	char *ptr = (char*) ilist->values;
	return (void*) &ptr[ilist->type->sizeoftype * itemno];
}

static inline void* ilist_add(ilist_t *ilist) {

	if (ilist->popped) {
		// we have popped stuff, so we cannot overwrite it!
		return NULL;
	}
	
	ilist->len++;
	if (ilist->len >= ilist->cap) {
		ilist->cap = ilist->cap * 2;
		ilist->values = mem_realloc_n(ilist->cap, ilist->type, ilist->values);
	}
	return ilist_get_internal(ilist, ilist->len - 1);
}

static inline void* ilist_pop(ilist_t *ilist) {

	if (ilist->len > 0) {
		ilist->popped = 1;
		ilist->len --;
		return ilist_get_internal(ilist, ilist->len);
	} else {
		// TODO: Error
		return NULL;
	}
}

static inline void* ilist_get(const ilist_t *ilist, int itemno) {

	if (ilist->len <= itemno) {
		return NULL;
	}
	return ilist_get_internal(ilist, itemno);
}

static inline void* ilist_last(const ilist_t *ilist) {

	if (ilist->len == 0) {
		return NULL;
	}
	return ilist_get_internal(ilist, ilist->len - 1);
}

#endif

