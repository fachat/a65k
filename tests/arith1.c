

#include <stdio.h>

#include "inline_list.h"
#include "tokenizer.h"
#include "print.h"


void print(const char *val) {

	do_print("---> %s", val);

	tokenizer_t *tok = tokenizer_create(val, 1);

	const ilist_t *out = NULL;

	arith_parse(tok, 0, &out);

	print_debug_arith(out);
	
	do_print("   > next is type %c, %s", tok->type, tok->line + tok->ptr);
}

int main(int argc, char *argv[]) {

	tokenizer_module_init();

	print("<1234");
	print("<$1234");
	print("<$x234");
	print("0x234");

	print("123+234");
	print("123+34*44");

	print("(123+34)*44");

}



