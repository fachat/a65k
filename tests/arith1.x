WRN:A:len=1
WRN:type=v, modifier=60 (<), op=0
