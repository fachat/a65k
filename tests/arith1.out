WRN:---> <1234
WRN:  A:len=1
WRN:    type=v, modifier=60 (<), op=0( ), val=(d)1234/4d2
WRN:   > next is type x, 
WRN:---> <$1234
WRN:  A:len=1
WRN:    type=v, modifier=60 (<), op=0( ), val=(h)4660/1234
WRN:   > next is type x, 
WRN:---> <$x234
WRN:  A:len=0
WRN:   > next is type e, $x234
WRN:---> 0x234
WRN:  A:len=1
WRN:    type=v, modifier=0 ( ), op=0( ), val=(H)564/234
WRN:   > next is type x, 
WRN:---> 123+234
WRN:  A:len=2
WRN:    type=v, modifier=0 ( ), op=0( ), val=(d)123/7b
WRN:    type=v, modifier=0 ( ), op=43(+), val=(d)234/ea
WRN:   > next is type x, 
WRN:---> 123+34*44
WRN:  A:len=3
WRN:    type=v, modifier=0 ( ), op=0( ), val=(d)123/7b
WRN:    type=v, modifier=0 ( ), op=43(+), val=(d)34/22
WRN:    type=v, modifier=0 ( ), op=42(*), val=(d)44/2c
WRN:   > next is type x, 
WRN:---> (123+34)*44
WRN:  A:len=2
WRN:    type=b, modifier=0( ), op=0( ) (
WRN:    A:len=2
WRN:      type=v, modifier=0 ( ), op=0( ), val=(d)123/7b
WRN:      type=v, modifier=0 ( ), op=43(+), val=(d)34/22
WRN:    type=v, modifier=0 ( ), op=42(*), val=(d)44/2c
WRN:   > next is type x, 
